### ISSUE TITLE
defer not working when navigating to the same route with different url search/params

### ISSUE THREAD

**Issue by NavaceSystem** [remix-run/remix#6637]
### What version of Remix are you using?

1.17

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Here's a quick repro: https://codesandbox.io/p/sandbox/smoosh-pine-7kwtrn?file=%2Fapp%2Froutes%2Ftest.%24delay.stuff.tsx%3A21%2C1

1- On initial load we get a `waiting for deferredThing` before the deferred data finishes.
2- Click `Navigate somewhere` to navigate (using `useNavigate`) to the same url but with a random value between 1 and 4 for the `delay` param.

### Expected Behavior

The url and UI should change immediately to reflect the new `delay` param and then render `waiting for deferredThing` exactly as it did on initial load.

### Actual Behavior

It waits for `deferredThing` to resolve before changing the url and UI.

From the docs here: https://remix.run/docs/en/main/guides/streaming#using-defer it says: 

> You will get a fallback rendered when the user navigates to the same route with different params (in the context of our above example, if the user selects from a list of packages on the left to find their location on the right).

So it sounds like defer behaviour should be working when navigating.

**Comment by brophdawg11** [comment]
We may need a docs update, but I think this is the expected React behavior for `startTransition` which was introduced in Remix 1.17 - React does not re-show fallbacks that have already resolved during transitions so if you want the `Suspense` boundary to  re-show you can add a `key` that produces a new boundary based on the param:

```jsx
<Suspense fallback={"waiting for deferredThing"} key={loaderData.delay}>
```

**Comment by NavaceSystem** [comment]
> We may need a docs update, but I think this is the expected React behavior for `startTransition` which was introduced in Remix 1.17 - React does not re-show fallbacks that have already resolved during transitions so if you want the `Suspense` boundary to re-show you can add a `key` that produces a new boundary based on the param:
> 
> ```js
> <Suspense fallback={"waiting for deferredThing"} key={loaderData.delay}>
> ```

I just tried this but it's not working because the url isn't updating till the deferred promise is resolved so UI is still blocked regardless of the key passed to Suspense. I can see the key to `Math.random()` but still the same behaviour.

**Comment by brophdawg11** [comment]
> the url isn't updating till the deferred promise is resolved 

I think this is just code sandbox not supporting streaming.  If I download your updated sandbox with `key={Math.random()}` and run it locally I do not see this behavior and the URL and fallback update immediately.

**Comment by NavaceSystem** [comment]
> > the url isn't updating till the deferred promise is resolved
> 
> I think this is just code sandbox not supporting streaming. If I download your updated sandbox with `key={Math.random()}` and run it locally I do not see this behavior and the URL and fallback update immediately.

I've tried it locally to on my real project first, the sandbox was just a tiny repro, and it's doing the same thing. I'm basically using the url as the key for the Suspense component but it doesn't do anything because the url changes only after the promise is resolved.

I'm basically trying to create a nice filtering experiencing while staying true to Remix i.e no client side fetching libraries.

In the example below the central loading spinner is being deferred which works on the initial page load. On subsequent page loads you can see that I don't see a loading spinner, I've got a global loader at the top of the page which starts spinning when I click on a different tab and the UI and URL only update after the deferred promise is resolved (the global loader stops).

https://github.com/remix-run/remix/assets/18566840/a7cae792-0b69-4a79-a833-f479fb07967c

**Comment by NavaceSystem** [comment]
> Ah - so in your example you're using `URLSearchParams`, not Dynamic Route Parameters (`/:param`). The note in the docs is referring to route parameters:
> 
> > You will get a fallback rendered when the user navigates to the same route with different params (in the context of our above example, if the user selects from a list of packages on the left to find their location on the right).
> 
> So I think at the moment this is "working as expected" since we specifically coded this to only re-fallback if the route parameters changed. Is there a reason you couldn't use `useNavigation` to show the spinner on subsequent navigations?
> 
> I can see the desire for defer to re-fallback in this specific example, but I think there are also plenty of cases where we wouldn't want it to re-fallback if a search param changed. We also wrote the original fallback logic before `startTransition` had been adopted so that may change some of the discussion. I'll loop back with the team and see what makes the most sense here.

Ah yes, you're right! I just played with one of the filters that change params and it does in fact show spinner correctly.

It's not just about the spinner, in the video the highlighting of the tab is based on the search param so it's a pretty poor experience for it to stay on the old tab for a few seconds after being clicked.

I can live with it for now, thanks so much for the help. It would be great if we opt-in to re-fallback based on search params.

**Comment by ncarrillo** [comment]
Ran into this same issue. Ended up going with useNavigation. 

I similarly had the same selected tab limitation, where it was based on some search parameter, and I solved for it using a bit of optimistic UI.

**Comment by brophdawg11** [comment]
> it's a pretty poor experience for it to stay on the old tab for a few seconds after being clicked.

What you're looking for is [Optimistic UI](https://remix.run/docs/en/main/guides/optimistic-ui).  You can choose to eagerly highlight the new tab during the navigation and even show the spinner instead of the content if you'd like.

**Comment by NavaceSystem** [comment]
Gave useNavigation a try and it's doing what I need. Thanks you.

**Comment by adarsh45** [comment]
I ran into the same issue
Situation: I was showing a search box inside a left menubar, and upon submitting that search form, it navigated user to a new URL from wherever the user is (left menubar is constant throughout the site)
Upon the first search hit, it navigates to my page, defers the loader return and shows Suspense fallback UI, but when that is over, when user wants to search one more thing
The page stays still and changes its render after the defer promise is fulfilled.
(Suspense fallback and key passing did not worked)

Solution:
If a dev wants to reshow the suspense fallback UI, then it must be the case that user is still on that previously defer loaded page.
This means your useNavigation() hooks state can come into play here!
and then it was easy ride from here, I used state and changed my UI again to the same UI that was in my Suspense fallback.

`const { state } = useNavigation();`

`
<RootContainer>
      {state !== "idle" ? (
        <SpinnerContainer />
      ) : (
        <Suspense fallback={<SpinnerContainer />}>
          <Await resolve={data} errorElement={<ErrorContainer />}>
            {(searchResults = []) => (
              <div className="pb-8">
                <DataTable
                  columns={columns}
                  rows={searchResults}
                />
              </div>
            )}
          </Await>
        </Suspense>
      )}
    </RootContainer>
`

**Comment by MoSattler** [comment]
hi @brophdawg11 - useNavigation will work in many case, but not in all. Why I can show a spinner this way when updating query params, this spinner will also show when navigating to another page, and that might not be the desired behaviour.

**Comment by MoSattler** [comment]
Hey there - just wanted to check in and see if there’s been any update or further thinking on the `defer` + URLSearchParams behavior. Appreciate all the insights shared so far.

**Comment by brophdawg11** [comment]
`defer` is deprecated once you adopt `v3_singleFetch` and is gone in React Router v7 when Single Fetch becoes the default.  Single Fetch does not have the same "await all defer promises on search params changes" that defer had so it should no longer be an issue there.
